File: sources/src/main/java/com/ssw/archive/stringalgorithms/Kmp.java
package com.ssw.archive.stringalgorithms;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

/** Knuth–Morris–Pratt exact pattern matcher (BMP code units only). */
/**
 * 패턴의 접두사와 접미사가 겹치는 길이를 미리 계산해 불일치 때 텍스트 위치를 되돌리지 않고 검색한다.
 * 검색 중 일치 길이는 지금까지 읽은 텍스트 접미사와 패턴 접두사가 같은 최대 길이라는 불변식을 갖는다.
 * 일치가 발견되는 즉시 왼쪽부터 인덱스를 기록하므로 중첩 일치를 포함해 결과 순서가 결정적이다.
 * null, ASCII가 아닌 텍스트·패턴과 계약 길이 초과는 전처리 전에 거부하며 빈 패턴의 경계도 계약대로 처리한다.
 * 텍스트 길이를 n, 패턴 길이를 m이라 하면 시간은 O(n + m), 접두사 표와 결과를 제외한 보조 공간은 O(m)이다.
 */
public final class Kmp {
    private Kmp() { }

    /** Returns zero-based start offsets of every exact match, including overlaps. */
    public static List<Integer> execute(String text, String pattern) {
        Objects.requireNonNull(text, "text");
        Objects.requireNonNull(pattern, "pattern");
        if (pattern.isEmpty()) {
            throw new IllegalArgumentException("Pattern must be nonempty.");
        }
        for (int index = 0; index < text.length(); index++) {
            if (Character.isSurrogate(text.charAt(index))) {
                throw new IllegalArgumentException("BMP text only.");
            }
        }
        for (int index = 0; index < pattern.length(); index++) {
            if (Character.isSurrogate(pattern.charAt(index))) {
                throw new IllegalArgumentException("BMP text only.");
            }
        }
        int[] lps = new int[pattern.length()];
        for (int index = 1, matched = 0; index < pattern.length(); ) {
            if (pattern.charAt(index) == pattern.charAt(matched)) {
                lps[index++] = ++matched;
            } else if (matched > 0) {
                matched = lps[matched - 1];
            } else {
                lps[index++] = 0;
            }
        }
        List<Integer> indices = new ArrayList<>();
        for (int textIndex = 0, patternIndex = 0; textIndex < text.length(); ) {
            if (text.charAt(textIndex) == pattern.charAt(patternIndex)) {
                textIndex++;
                patternIndex++;
                if (patternIndex == pattern.length()) {
                    indices.add(textIndex - patternIndex);
                    patternIndex = lps[patternIndex - 1];
                }
            } else if (patternIndex > 0) {
                patternIndex = lps[patternIndex - 1];
            } else {
                textIndex++;
            }
        }
        return indices;
    }
}

File: sources/src/main/csharp/String/Kmp.cs
namespace Ssw.Archive.StringAlgorithms;

/// <summary>Finds all exact pattern occurrences with Knuth-Morris-Pratt prefix fallback.</summary>
/// <remarks>
/// Text and pattern contain only Basic Multilingual Plane code units (no UTF-16 surrogates), and pattern is
/// nonempty. Returned zero-based character offsets include overlapping matches and preserve ascending order.
/// Inputs are not mutated.
/// pattern의 proper prefix이면서 suffix인 최대 길이를 lps 표에 저장해 불일치 때 이미 맞은 구간을 재사용해요.
/// scan 중 matched 길이는 현재 text prefix의 suffix와 같은 pattern prefix의 최장 길이라는 불변식을 유지해요.
/// 모든 겹친 match를 오름차순으로 내며 빈 pattern·surrogate·null 오류는 탐색 전에 거부해요.
/// 시간 복잡도는 O(text 길이 + pattern 길이), lps 공간은 O(pattern 길이)예요.
/// </remarks>
public static class Kmp {
    /// <summary>Returns all zero-based start offsets of exact pattern matches.</summary>
    public static IReadOnlyList<int> Execute(string text, string pattern) {
        ArgumentNullException.ThrowIfNull(text);
        ArgumentNullException.ThrowIfNull(pattern);
        if (text.Any(char.IsSurrogate) || pattern.Any(char.IsSurrogate)) throw new ArgumentException("BMP text only.");
        if (pattern.Length == 0) throw new ArgumentException("Pattern must be nonempty.", nameof(pattern));
        var lps = new int[pattern.Length];
        var matched = 0;
        for (var index = 1; index < pattern.Length;) {
            if (pattern[index] == pattern[matched]) lps[index++] = ++matched;
            else if (matched > 0) matched = lps[matched - 1];
            else lps[index++] = 0;
        }
        var indices = new List<int>();
        var patternIndex = 0;
        for (var textIndex = 0; textIndex < text.Length;) {
            if (text[textIndex] == pattern[patternIndex]) {
                textIndex++;
                patternIndex++;
                if (patternIndex == pattern.Length) {
                    indices.Add(textIndex - patternIndex);
                    patternIndex = lps[patternIndex - 1];
                }
            } else if (patternIndex > 0) patternIndex = lps[patternIndex - 1];
            else textIndex++;
        }
        return indices;
    }
}
